planner/core/logical_plan_builder.go

```diff
@@ -2382,25 +2374,23 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
 	return projUponView, nil
 }
 
-// projectVirtualColumns is only for DataSource. If some table has virtual generated columns,
-// we add a projection on the original DataSource, and calculate those columns in the projection
-// so that plans above it can reference generated columns by their name.
-func (b *PlanBuilder) projectVirtualColumns(ctx context.Context, ds *DataSource, columns []*table.Column) (*LogicalProjection, error) {
-	hasVirtualGeneratedColumn := false
+// prepareVirtualColumns is only for DataSource.
+// It prepares virtualColExprs and virtualColSchema for table which has virtual generated columns.
+// virtualColExprs and virtualColSchema are used to rewrite virtual columns in pushDownSelAndResolveVirtualCols.
+func (b *PlanBuilder) prepareVirtualColumns(ctx context.Context, ds *DataSource, columns []*table.Column) error {
+	hasVirtualCol := false
 	for _, column := range columns {
 		if column.IsGenerated() && !column.GeneratedStored {
-			hasVirtualGeneratedColumn = true
+			hasVirtualCol = true
 			break
 		}
 	}
-	if !hasVirtualGeneratedColumn {
-		return nil, nil
+	if !hasVirtualCol {
+		return nil
 	}
-	proj := LogicalProjection{
-		Exprs:            make([]expression.Expression, 0, len(columns)),
-		calculateGenCols: true,
-	}.Init(b.ctx)
 
+	ds.virtualColSchema = ds.Schema().Clone()
```

Comment:
Don't clone it is okay?

Reply:
It may be modified by `planner` when doing logical optimizations, so I use `Clone()` for safety.
And only queries with virtual columns will call this clone method, so it's okay to use it in these cases?